clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +347,537 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+     */
+    @Override
+    public boolean shouldHeartbeatNow() {
+        MemberState state = state();
+        return state == MemberState.ACKNOWLEDGING || state == MemberState.LEAVING;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onHeartbeatRequestSent() {
+        MemberState state = state();
+        if (state == MemberState.ACKNOWLEDGING) {
+            if (allPendingAssignmentsReconciled()) {
+                transitionTo(MemberState.STABLE);
+            } else {
+                transitionTo(MemberState.RECONCILING);
+            }
+        } else if (state == MemberState.LEAVING) {
+            transitionTo(MemberState.UNSUBSCRIBED);
         }
-        return state.equals(MemberState.STABLE);
     }
 
     /**
-     * Take new target assignment received from the server and set it as targetAssignment to be
-     * processed. Following the consumer group protocol, the server won't send a new target
-     * member while a previous one hasn't been acknowledged by the member, so this will fail
-     * if a target assignment already exists.
+     * @return True if there are no assignments waiting to be resolved from metadata or reconciled.
+     */
+    private boolean allPendingAssignmentsReconciled() {
+        return assignmentUnresolved.isEmpty() && assignmentReadyToReconcile.isEmpty();
+    }
+
+    @Override
+    public boolean shouldSkipHeartbeat() {
+        MemberState state = state();
+        return state == MemberState.UNSUBSCRIBED || state == MemberState.FATAL;
+    }
+
+    /**
+     * Reconcile the assignment that has been received from the server and for which topic names
+     * are resolved, kept in the {@link #assignmentReadyToReconcile}. This will commit if needed,
+     * trigger the callbacks and update the subscription state. Note that only one reconciliation
+     * can be in progress at a time. If there is already another one in progress when this is
+     * triggered, it will be no-op, and the assignment will be reconciled on the next
+     * reconciliation loop.
+     */
+    boolean reconcile() {
+        // Make copy of the assignment to reconcile as it could change as new assignments or metadata updates are received
+        SortedSet<TopicPartition> assignedPartitions = new TreeSet<>(assignmentReadyToReconcile);
+        boolean sameAssignmentReceived = assignedPartitions.equals(subscriptions.assignedPartitions());
+
+        if (reconciliationInProgress || sameAssignmentReceived) {
+            String reason;
+            if (reconciliationInProgress) {
+                reason = "Another reconciliation is already in progress. Assignment " +
+                        assignmentReadyToReconcile + " will be handled in the next reconciliation loop.";
+            } else {
+                reason = "Target assignment ready to reconcile is equals to the member current assignment.";
+            }
+            log.debug("Ignoring reconciliation attempt." + reason);
```

Comment:
```suggestion
            log.debug("Ignoring reconciliation attempt. " + reason);
```

Nit: it'll be visually easier to parse with the space before the next sentence.